src/librustc_typeck/check/mod.rs

```diff
@@ -1801,32 +1801,72 @@ fn check_packed_inner<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     false
 }
 
+fn bad_variant_count<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, adt: &'tcx ty::AdtDef, sp: Span) {
+    let variant_spans: Vec<_> = adt.variants.iter().map(|variant| {
+        tcx.hir().span_if_local(variant.def_id).unwrap()
+    }).collect();
+    let msg = format!(
+        "needs exactly one variant, but has {}",
+        adt.variants.len(),
+    );
+    let mut err = struct_span_err!(tcx.sess, sp, E0731, "transparent enum {}", msg);
+    err.span_label(sp, &msg);
+    match &variant_spans[..] {
+        &[] => {},
+        &[ref start.., ref end] => {
+            for variant_span in start {
+                err.span_label(*variant_span, "");
+            }
+            err.span_label(*end, &format!(
+                "too many variants in `{}`",
+                tcx.def_path_str(def_id),
+            ));
+        },
+    }
+    err.emit();
+}
+
+fn bad_non_zero_sized_fields<'a, 'tcx, I: Iterator<Item = Span>>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    field_spans: I,
+    sp: Span,
+) {
+    let msg = format!("needs exactly one non-zero-sized field, but has {}", non_zst_count);
+    let mut err = struct_span_err!(
+        tcx.sess,
+        sp,
+        E0690,
+        "{}transparent {} {}",
+        if adt.is_enum() { "the variant of a " } else { "" },
+        adt.descr(),
+        msg,
+    );
+    err.span_label(sp, &msg);
+    for sp in field_spans {
+        err.span_label(sp, "this field is non-zero-sized");
```

Comment:
...and this nicely by having `.span_labels(field_spans, "this field is non-zero-sized")` so that you can chain and don't have to add `let mut err = ...;`

Reply:
Wow, this method has been there since 2016 and I hadn't noticed... 😊

Reply:
...but you can add a new method that takes an iterator of spans?

Reply:
@Centril I don't think it'll be as general as it looks: we have cases where we give every span the same label, cases where we only give a label to the first span, cases where we only give the last span a label and cases where we give each span in the iterator a different label using `format!`. I would prefer such a method to be added with more foresight in a concerted effort to clean up the codebase than as a one off in a random PR.

Reply:
@rust-lang/wg-diagnostics ^^